power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp

```diff
@@ -307,9 +288,11 @@ class MainModelImpl {
     void update_state(UpdateChange const& changes) {
         // if topology changed, everything is not up to date
         // if only param changed, set param to not up to date
-        is_topology_up_to_date_ = is_topology_up_to_date_ && !changes.topo;
-        is_sym_parameter_up_to_date_ = is_sym_parameter_up_to_date_ && !changes.topo && !changes.param;
-        is_asym_parameter_up_to_date_ = is_asym_parameter_up_to_date_ && !changes.topo && !changes.param;
+        solvers_cache_status.set_topology_status(solvers_cache_status.is_topology_valid() && !changes.topo);
+        solvers_cache_status.template set_parameter_status<symmetric_t>(
+            solvers_cache_status.template is_parameter_valid<symmetric_t>() && !changes.topo && !changes.param);
+        solvers_cache_status.template set_parameter_status<asymmetric_t>(
+            solvers_cache_status.template is_parameter_valid<asymmetric_t>() && !changes.topo && !changes.param);
```

Comment:
this function feels less readable in its current form. would it be possible to refactor?

to me this seems like adding separate `set` and `reset` functions actually makes sense... That way, we don't accidentally set anything to `true` here. e.g.

```suggestion
        auto const invalid_topo = changes.topo;
        auto const invalid_params = invalid_topo || changes.param;

        solvers_cache_status.invalidate_topology(invalid_topo);
        solvers_cache_status.template invalidate_parameter<symmetric_t>(invalid_params);
        solvers_cache_status.template invalidate_parameter<asymmetric_t>(invalid_params);
```

or

```suggestion
        solvers_cache_status.change_topology(changes.topo);
        solvers_cache_status.template parameters_changed<symmetric_t>(changes.topo || changes.param);
        solvers_cache_status.template parameters_changed<asymmetric_t>(changes.topo || changes.param);
```

and separately, maybe this function is not in the correct location in the first place

```suggestion
        solvers_cache_status.apply(changes);
```

still not perfect but better than what it is now...